content/blog/2020-05-26-may-20-community-gems.md

```diff
@@ -0,0 +1,93 @@
+---
+title: May '20 Community Gems
+date: 2020-05-26
+description: |
+  A roundup of technical Q&A's from the DVC community. This month, we discuss 
+  development best practices, sharing models and data across projects, 
+  and using DVC with teams.
+descriptionLong: |
+  A roundup of technical Q&A's from the DVC community. This month, we discuss 
+  development best practices, sharing models and data across projects, 
+  and using DVC with teams.
+picture: 2020-05-26/May_20_Gems_Header.png
+author: elle_obrien
+commentsUrl: https://discuss.dvc.org/t/april-20-community-gems/356
+tags:
+  - Discord
+  - Gems
+  - Cache
+  - Google Cloud Storage
+  - Import
+---
+
+## Discord gems
+
+Here are some Q&A's from our Discord channel that we think are worth sharing.
+
+### Q: [How do I completely delete a file from DVC?](https://discord.com/channels/485586884165107732/563406153334128681/710546561498873886)
```

Comment:
it's a good one and very valuable! there are some caveats to this, though:

1. removing a file from dir - not as simple, it would be nice to mention or even cover this as well to make the A especially useful.
2. when we remove manually a file from cache there is a risk that it was referenced from another commit (possibly even under a different name). `gc --all-commits` handles this safely.

let me know if you need my help with this one - we can discuss the details.

Reply:
i think i need a little help understanding the first point- why is removing a `.dvc` file from a directory not so simple? the second point seems reasonable

Reply:
so, regarding the first point:

There is a level (file) of indirection in the way DVC stores directories. It solves two problems (that you would immediately hit when you implement DVC itself):

1. For a file we have a simple one-to-one relationship - file to its hash (`md5` in our case). We need this hash to find and save it in DVC cache/remote. We need it to being able to check if we need to run a stage again (if there are some changes in a dependency), etc. DVC saves hash value in its YAML files a one line: `md5: 12324dedefdsfd`. But what about dirs? How do we get a simple _hash_ for the whole directory? We can't store 1M md5s into a `.dvc` file.

2. Directories are hierarchical - they have subdirectories, they have a lot of files with different names inside. How do we capture all these file names? All this complex structure? Again, we can't save it into a `.dvc` file - too much info for a simple file, it can be megabytes in size, etc.

So, to address those two problems, in DVC we store a special JSON file `122sdsd323e42.dir` with a content of the directory. You can read a bit about this here  - https://dvc.org/doc/user-guide/dvc-files-and-directories#structure-of-cache-directory .

It means that if I want to remove `data/path/file-a.txt`, I would need to find it's cache first by parsing the `.dir` files that corresponds to the `data`.

Please, let me know if it's more or less makes sense :)

Reply:
oh, i see- if the file was in a directory that had been DVC tracked at a directory level. Got it, thanks! I think i can answer it now

Reply:
@shcheklein just gave writing the answer a try. let me know what you think!

Reply:
It's close! :) One correction - when `.dir` is located we can't edit this file (it's name is md5 of the content and we need to preserve it). The only things we can do is to find an entry inside, that entry will have md5 on its own and then delete file in the way described above. The caveat here is that DVC will rely on the information in `.dir` in certain cases. E.g. it will never push a file to remote storage again.

Btw, we don't mention remote storage here at all - it's another aspect of getting rid of data in DVC.

Reply:
aha! ok, fixed that and pushed :D